cookbooks/bach_krb5/recipes/keytabs.rb

```diff
@@ -0,0 +1,42 @@
+get_all_nodes().each do |h|
+  # Create directories to store keytabs
+  directory "#{node[:bcpc][:hadoop][:kerberos][:keytab][:dir]}/#{float_host(h[:fqdn])}" do
+    action :create
+    owner "root"
+    group "root"
+    mode 0755
+    recursive true
+  end
+  
+  # Generate all the principals
+  node[:bcpc][:hadoop][:kerberos][:data].each do |ke, va|
+
+    krb5_principal "#{va['principal']}/#{float_host(h[:fqdn])}@#{node[:bcpc][:hadoop][:realm]}" do
+      action :delete
+      only_if {principal_exists?("#{va['principal']}/#{float_host(h[:fqdn])}@#{node[:bcpc][:hadoop][:realm]}") && node[:bcpc][:hadoop][:kerberos][:keytab][:recreate] == true}
+    end
+
+    krb5_principal "#{va['principal']}/#{float_host(h[:fqdn])}@#{node[:bcpc][:hadoop][:realm]}" do
+      action :create
+      randkey true
+      not_if {principal_exists?("#{va['principal']}/#{float_host(h[:fqdn])}@#{node[:bcpc][:hadoop][:realm]}")}
+    end
+  end
+
+  # Generate keytabs
+  node[:bcpc][:hadoop][:kerberos][:data].each do |ke, va|
```

Comment:
Nit: could `ke` and `va` be more descriptive? Or the relevant data needed out of `va` be stored into a descriptive variable name to make it obvious what is actually at use below without lots of reading. (It is just using the keytab file name right?)